src_kt/java/arcs/util/Guard.kt

```diff
@@ -0,0 +1,85 @@
+package arcs.arcs.util
+
+import kotlinx.coroutines.sync.Mutex
+import kotlin.properties.ReadWriteProperty
+import kotlin.reflect.KProperty
+
+/**
+ * Builds a [Guard] property delegate where all access/mutation to the delegated property must be
+ * done within a lock on the provided [mutex].
+ *
+ * [initialValue] is a function which returns the starting value of the property.
+ *
+ * Example:
+ *
+ * ```kotlin
+ * class MyClass {
+ *     val mutex = Mutex()
+ *     var myProtectedString: String by guardWith(mutex) { "Hello world." }
+ *
+ *     suspend fun succeed() {
+ *         val randomNum = Random.nextInt()
+ *         mutex.withLock { myProtectedString = "Random: $randomNum" }
+ *     }
+ *
+ *     suspend fun crash() {
+ *         val randomNum = Random.nextInt()
+ *         myProtectedString = "Random: $randomNum"
+ *     }
+ * }
+ * ```
+ */
```

Comment:
Is there any magical way to also have this bestow an @GuardedBy("mutex") annotation on the field so that error-prone will come time check it? Seems we'd need something more power than reified, akin to a macro.

Reply:
Unfortunately not :(